pandas/io/tests/test_excel.py

```diff
@@ -518,6 +516,125 @@ def test_reader_seconds(self):
+        with ensure_clean('.xlsx') as pth:
+            for c_idx_names in [True, False]:
+                for r_idx_names in [True, False]:
+                    for c_idx_levels in [1, 3]:
+                        for r_idx_levels in [1, 3]:
+                            # column index name can't be serialized unless MultiIndex
+                            if (c_idx_levels == 1 and c_idx_names):
+                                continue
+
+                            # empty name case current read in as unamed levels, not Nones
+                            check_names = True
+                            if not r_idx_names and r_idx_levels > 1:
+                                check_names = False
+
+                            df = mkdf(5, 5, c_idx_names,
+                                        r_idx_names, c_idx_levels,
+                                        r_idx_levels)
+                            df.to_excel(pth)
+                            act = pd.read_excel(pth, index_col=list(range(r_idx_levels)),
+                                                header=list(range(c_idx_levels)))
+                            tm.assert_frame_equal(df, act, check_names=check_names)
+
+                            df.iloc[0, :] = np.nan
+                            df.to_excel(pth)
+                            act = pd.read_excel(pth, index_col=list(range(r_idx_levels)),
+                                                header=list(range(c_idx_levels)))
+                            tm.assert_frame_equal(df, act, check_names=check_names)
+
+                            df.iloc[-1, :] = np.nan
+                            df.to_excel(pth)
+                            act = pd.read_excel(pth, index_col=list(range(r_idx_levels)),
+                                                header=list(range(c_idx_levels)))
+                            tm.assert_frame_equal(df, act, check_names=check_names)
+
+    def test_excel_oldindex_format(self):
+        #GH 4679
+        data = np.array([['R0C0', 'R0C1', 'R0C2', 'R0C3', 'R0C4'],
+                         ['R1C0', 'R1C1', 'R1C2', 'R1C3', 'R1C4'],
+                         ['R2C0', 'R2C1', 'R2C2', 'R2C3', 'R2C4'],
+                         ['R3C0', 'R3C1', 'R3C2', 'R3C3', 'R3C4'],
+                         ['R4C0', 'R4C1', 'R4C2', 'R4C3', 'R4C4']])
+        columns = ['C_l0_g0', 'C_l0_g1', 'C_l0_g2', 'C_l0_g3', 'C_l0_g4']
+        mi = MultiIndex(levels=[['R_l0_g0', 'R_l0_g1', 'R_l0_g2', 'R_l0_g3', 'R_l0_g4'],
+                                ['R_l1_g0', 'R_l1_g1', 'R_l1_g2', 'R_l1_g3', 'R_l1_g4']],
+                        labels=[[0, 1, 2, 3, 4], [0, 1, 2, 3, 4]],
+                        names=['R0', 'R1'])
+        si = Index(['R_l0_g0', 'R_l0_g1', 'R_l0_g2', 'R_l0_g3', 'R_l0_g4'],  name='R0')
+
+        in_file = os.path.join(self.dirpath, 'test_index_name_pre17' + self.ext)
+
+        expected = pd.DataFrame(data, index=si, columns=columns)
+        with tm.assert_produces_warning(FutureWarning):
```

Comment:
add a test where you are passing ``has_index_names=False`` (and assert that it warns as well)